File: src/leetcode/python/AddTwoNumbers_2.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class AddTwoNumbers_2(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        plus = 0
        head = l2
        while l1 and l2:
            val = l1.val + l2.val + plus
            l2.val = val % 10
            plus = val / 10
            if l1.next is None or l2.next is None:
                break
            l1, l2 = l1.next, l2.next
        while l1.next:
            l1 = l1.next
            val = l1.val + plus
            l2.next = ListNode(val % 10)
            plus = val / 10
            l2 = l2.next
        while l2.next:
            l2 = l2.next
            val = l2.val + plus
            l2.val = val % 10
            plus = val / 10
        if plus:
            l2.next = ListNode(1)
        return head

File: src/leetcode/java/AddTwoNumbers_2.java
package leetcode.java;

public class AddTwoNumbers_2 {
  public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
    ListNode head = l2;
    int plus = 0;
    int val = 0;
    while (l1 != null && l2 != null) {
      val = l1.val + l2.val + plus;
      l2.val = val % 10;
      plus = val / 10;
      if (l1.next == null || l2.next == null) break;
      l1 = l1.next;
      l2 = l2.next;
    }
    while (l1.next != null) {
      l1 = l1.next;
      val = l1.val + plus;
      plus = val / 10; 
      l2.next = new ListNode(val % 10);
      l2 = l2.next;
    }
    while (l2.next != null) {
      l2 = l2.next;
      val = l2.val + plus;
      plus = val / 10;
      l2.val = val % 10;
    }
    if (plus > 0) l2.next = new ListNode(1);
    return head;
  }

  class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
      val = x;
    }
  }
}
